### DISCUSSION TITLE
A NavigationService for .NET MAUI

### DISCUSSION THREAD

**Discussion by luismts** [PrismLibrary/Prism#2725] [.NET MAUI]
Hi guys,

I was thinking about a NavigationService for .NET MAUI and I have tested a lot of things trying to have something similar to Prism. In the end, because Prism is OpenSource, I have copied its navigation. You can see the repo [HERE](https://github.com/mallaagency/NavigationService).

Also, I have shared some progress on Twitter and I'm really impressed by two things

1. The reaction of some people
2. The support of others

You can see the thread [here](https://twitter.com/luismatosluna/status/1554850418861146115).

> Sorry for not taking this initiative before but hey I'm here. That's what counts, right?

### The idea

Have a NavigationService for .NET MAUI but without any other services and using default container management that .NET MAUI has. Many of the navigation features remain largely untouched. As part of this rewrite, some attention to making it easier for everyone to use.

>As a general resume, you will have Prism.NavigationService + CommunityToolkit = NOT NEED FRAMEWORKS
>At least for simple projects. Not all projects need all the features.

### Why?

Maui presents the app startup process paradigm that uses the AppHostBuilder pattern you may be familiar with from AspNetCore and other .NET Core applications. So, with this we can use dependency injection in our apps; having this, we dont need a particular framework for MVVM because .NET MAUI + CommunityToolkit has all the tools we need.

For example, there are some remarkable MVVM frameworks, but once we have the NavigationService, it doesn't make sense to use them. Most of them are responsible for many services that we have today with the CommunityToolkit.

### Comparison

Icon | Description
-- | --
✅ | Done
💔 | Never implemented
🚫 | Not need it or Optional
👀 | Can be easily implemented

Feature | .NET MAUI | Community Toolkit | Prism
-------- | ------------ | --------------------- | --------
Commands | ✅ | ✅ | ✅
Dependency Injection | ✅ | 🚫 | ✅
Navigation | 💔 | 🚫 | ✅
ViewModelLocator | 💔 | 🚫 | ✅
Dialog Service | 👀 | ✅ | ✅
Behaviors | 👀 | ✅ | ✅
Event Aggregation | 🚫 | 🚫 | ✅
Modules | 🚫 | 🚫 | ✅
Regions | 🚫 | 🚫 | ✅

**Note**: Navigation support 

1. Default navigation feature 
2. Page life cycle on ViewModel
3. XAML navigation 

### Conclusion

I know some people would like this because of the support and also know there are people that want the complete framework. So, giving us the flexibility to choose is nice.

So, seeing all the things shared here; **do you think this is something that can be done in the short or mid-time?**

**Comment by dansiegel** [comment]
What you've done is identify a non-issue. Prism today works fantastic with the Community Toolkit and .NET MAUI... and ReactiveUI. Using Prism does not require you to make your ViewModels inherit from Prism's BindableBase... there is no requirement that you use DelegateCommand... 

The reality is you can mix and match these libraries as is. You can make use of the CommunityToolkit while using Prism to provide you with the Navigation benefits.

What you've done in the project isn't to just copy our code, but it provides a broken implementation that does not set people up for success. I certainly welcome any community feedback on this, but as you have laid this issue out I do not see this as an issue that "needs to be solved", but rather a solution for a problem that never existed.
